rllib/tuned_examples/sac/pendulum_sac_envrunner.py

```diff
@@ -0,0 +1,59 @@
+import ray
+
+from ray import air, tune
+from ray.rllib.algorithms.sac.sac import SACConfig
+from ray.rllib.env.single_agent_env_runner import SingleAgentEnvRunner
+
+config = (
+    SACConfig()
+    # Enable new API stack and use EnvRunner.
+    .experimental(_enable_new_api_stack=True)
+    .rollouts(
+        rollout_fragment_length=1,
+        env_runner_cls=SingleAgentEnvRunner,
+        num_rollout_workers=1,
+    )
+    .environment(env="Pendulum-v1")
+    .training(
+        lr=3e-4,
+        target_entropy="auto",
+        # TODO (simon): Implement n-step.
```

Comment:
We should do this inside the episode buffer:

When calling `buffer.sample()`, the buffer would pick a timestep in some episode and do the n-step on the fly. This way we can:
* still only store each individual obs/action/reward (saves memory)
* support n-step with a range of n (picked at random)

Reply:
As far as I see it we only need to use `batch_length_T=self.config.n_step + 1` in `buffer.sample()` and drop all steps in between the first and the last then correct?

Reply:
We also need to add all the rewards in between.

I do think it's best to do n-step in the buffer. It's the natural location to do it (right after the raw/space-saving episode data is retrieved). If we did it before, we would have to store all possible combinations of n-step samples. If we did it afterwards (e.g. in training_step), we would send more data through the network (well, at least if the buffer were on another node, e.g. for APEX algos??).

But happy to hear your opinion or objections to this.

Reply:
Your arguments make fully sense. I was thinking to keep the buffer as clear as possible from algorithmic details. I was thinking about a `BufferToLearnerConnector` that does these kind of formations on the sampled data - this one could also run on the nodes and it would leave the buffer with some very basic things to do: store experiences and sample them (either shuffled or in sequences with or without prioritization). I see however, that the way from buffer to learner goes through the algorithm, so at least the naming of the connector would not be that favourable. 

The thing I see is that we cannot sample n-step from the buffer without sampling all experiences in the sequence and then drop the in-between ones after we have calculated rewards. And that would also work in a connector I guess. Another possibility is to provide buffers with connectors (like the `EnvRunner`)  and filter their samples through them. 

All of these are simply ideas - not fully thought through. But it could be in near future that we need customizable sampling from experience buffers.